packages/celotool/src/cmds/bots/auto-vote.ts

```diff
@@ -0,0 +1,318 @@
+  for (const vote of currentVotes) {
+    const revokeTxs = await election.revoke(botAccount, vote.group, vote.pending.plus(vote.active))
+    await concurrentMap(10, revokeTxs, (tx) => {
+      return tx.sendAndWaitForReceipt({ from: botAccount })
+    })
+    const oldCapacity = groupCapacities.get(vote.group)!
+    groupCapacities.set(vote.group, oldCapacity.plus(vote.pending.plus(vote.active)))
+  }
+
+  const groupCapacity = groupCapacities.get(voteForGroup)!
+  const voteAmount = BigNumber.minimum(lockedGoldAmount, groupCapacity)
+  const voteTx = await election.vote(voteForGroup, BigNumber.minimum(voteAmount))
+  await voteTx.sendAndWaitForReceipt({ from: botAccount })
+  console.info(`Completed voting as ${botAccount}`)
+
+  groupCapacities.set(voteForGroup, groupCapacity.minus(voteAmount))
+}
+
+// NOTE: commented out because this relies on an unpublished contract kit version
+// // After saturation, explore never-elected-Validator-Groups in FIFO order.
+// // @ts-ignore: declared but its value is never read.
+// async function getNextNeverElectedValidatorGroup(kit: ContractKit): Promise<Address> {
+//   const validators = await kit.contracts.getValidators()
+//   const groups = await validators.getRegisteredValidatorGroups()
+//   // Sort Validator Groups by max(ValidatorGroup.sizeHistory), e.g. last Validator.member update timestamp.
+//   groups.sort((a, b) => a.membersUpdated - b.membersUpdated)
+//   for (const group of groups) {
+//     const groupValidators: Validator[] = await concurrentMap(10, group.members, (member) =>
+//       validators.getValidator(member)
+//     )
+//     // If no member validator has a score then this group hasn't been elected since group.membersUpdated
+//     if (0 === groupValidators.reduce((a, b) => a.plus(b.score), new BigNumber(0)).toNumber()) {
+//       return group.address
+//     }
+//   }
+//   return NULL_ADDRESS
+// }
+
+async function calculateInitialGroupCapacities(kit: ContractKit): Promise<Map<string, BigNumber>> {
+  console.info('Determining which groups have capacity for more votes')
+
+  const validators = await kit.contracts.getValidators()
+  const election = await kit.contracts.getElection()
+
+  const groupCapacities = new Map<string, BigNumber>()
+  for (const groupAddress of await validators.getRegisteredValidatorGroupsAddresses()) {
+    const vgv = await election.getValidatorGroupVotes(groupAddress)
+    if (vgv.eligible) {
+      groupCapacities.set(groupAddress, vgv.capacity)
+    }
+  }
+
+  return groupCapacities
+}
+
+async function calculateGroupScores(kit: ContractKit): Promise<Map<string, BigNumber>> {
+  console.info('Calculating weights of groups based on the scores of elected validators')
+  const election = await kit.contracts.getElection()
+  const validators = await kit.contracts.getValidators()
+
+  const validatorSigners = await election.getCurrentValidatorSigners()
+  const validatorAccounts = (
```

Comment:
Could we use `validators.getRegisteredValidators()` here?  Do we want to include previously elected (but currently unelected) validators?

Reply:
As I understand it, the only validators that have a score are ones that are _currently_ elected. If I am mistaken about that, then sure, this can be changed. This function could probably remain intact otherwise.

Reply:
I think that when incumbent validators lose election, their validator score remains whatever it was for the last epoch they were elected.

```
celocli validator:list | grep 0\\. | wc -l
Fetching Validators... done
     120
```

Reply:
yeah, you're right. They seem to preserve the score from when they were last elected.

However, that, in itself, does not lead to a decision on what should happen with this line of code.